services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java

```diff
@@ -457,13 +458,16 @@ public Response export(UriInfo uriInfo, RealmModel realm, String format) {
             if (getConfig().isSignSpMetadata())
             {
                 KeyManager.ActiveRsaKey activeKey = session.keys().getActiveRsaKey(realm);
-                String keyName = getConfig().getXmlSigKeyInfoKeyNameTransformer().getKeyName(activeKey.getKid(), activeKey.getCertificate());
-                KeyPair keyPair = new KeyPair(activeKey.getPublicKey(), activeKey.getPrivateKey());
+                X509Certificate certificate = activeKey.getCertificate();
+                String keyName = getConfig().getXmlSigKeyInfoKeyNameTransformer().getKeyName(activeKey.getKid(), certificate);
+                // No need to expose pubkey separately if cert is available
+                KeyPair keyPair = new KeyPair(certificate == null ? activeKey.getPublicKey() : null, activeKey.getPrivateKey());
```

Comment:
Buf! This is very weird. I think the correct way of fixing this is passing the certificate in [the SAML2Signature class to the sing call](https://github.com/keycloak/keycloak/blob/21.0.1/saml-core/src/main/java/org/keycloak/saml/processing/api/saml/v2/sig/SAML2Signature.java#L144) (there is a variant in [XMLSignatureUtil.sign](https://github.com/keycloak/keycloak/blob/21.0.1/saml-core/src/main/java/org/keycloak/saml/processing/core/util/XMLSignatureUtil.java#L369-L370) which receives the certificate parameter). And in the same class just add to the KeyInfo the certificate or the Public Key, not both [here](https://github.com/keycloak/keycloak/blob/21.0.1/saml-core/src/main/java/org/keycloak/saml/processing/core/util/XMLSignatureUtil.java#L750-L756). This second part can be something like:

```java
        if (x509Certificate != null) {
            items.add(keyInfoFactory.newX509Data(Collections.singletonList(x509Certificate)));
        } else if (publicKey != null) {
            items.add(keyInfoFactory.newKeyValue(publicKey));
        }
```

I think this is fixing the issue in all the places. The X509Certificate is always sent if present, if not the PublicKey info. I have checked in several places and the behavior is now quite erratic (sometimes only the public key is added, sometimes both). This way we are doing always the same, preferring cert over key in all the calls.

WDYT?

Reply:
Thanks for the feedback! Indeed I had originally expected the `SAML2Signature` to prefer certs over public keys, and was surprised that it didn't. Being unsure if the behavior was intentional, I went for the limited workaround in `SAMLIdentityProvider`, but given your feedback, I'm more than happy to change `SAML2Signature` semantics instead.

Reply:
Thanks @dkobras! I also thought about being conservative at the beginning. But, after checking the behavior was quite erratic depending the code path followed (sometimes only KeyValue is added to the KeyInfo, sometimes both KeyValue and X509Data are added), I thought that doing always the same proper thing was better.

Reply:
@rmartinc Thanks! I've updated the PR based on your feedback. Please let me know if this is what you had in mind.